{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "U8_7byeLT8P2"
      },
      "outputs": [],
      "source": [
        "import requests\n",
        "from pathlib import Path\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#output directory\n",
        "DATA_DIR = Path(\"data/uniprot_sequences\")\n",
        "DATA_DIR.mkdir(parents=True, exist_ok=True)\n"
      ],
      "metadata": {
        "id": "n2B5yKRcUIE0"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#uniport url link to download large sequences\n",
        "UNIPROT_URL = \"https://rest.uniprot.org/uniprotkb/stream\"\n"
      ],
      "metadata": {
        "id": "2WRNbDdfUSFW"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#virus queries and taxonomy ids\n",
        "virus_queries = {\n",
        "    \"HIV\": \"taxonomy_id:11676\",\n",
        "    \"HCV\": \"taxonomy_id:11103\",\n",
        "    \"SARS_CoV_2\": \"taxonomy_id:2697049\"\n",
        "}"
      ],
      "metadata": {
        "id": "8TJn_l3wUucV"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def download_uniprot_fasta(virus_name, query, max_records=10000):\n",
        "    pass\n"
      ],
      "metadata": {
        "id": "VRFvcdqVU2-H"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def download_uniprot_fasta(virus_name, query, max_records=10000):\n",
        "    params = {\n",
        "        \"query\": query,\n",
        "        \"format\": \"fasta\",\n",
        "        \"size\": max_records\n",
        "    }\n",
        "    return params\n",
        "\n",
        "print(download_uniprot_fasta(\"HIV\", \"taxonomy_id:11676\"))\n",
        "print(download_uniprot_fasta(\"HCV\", \"taxonomy_id:11103\"))\n",
        "print(download_uniprot_fasta(\"SARS_CoV_2\", \"taxonomy_id:2697049\"))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "esTDhyOUVfOx",
        "outputId": "e3047003-bc07-45ee-c2d0-abf767531e7a"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "{'query': 'taxonomy_id:11676', 'format': 'fasta', 'size': 10000}\n",
            "{'query': 'taxonomy_id:11103', 'format': 'fasta', 'size': 10000}\n",
            "{'query': 'taxonomy_id:2697049', 'format': 'fasta', 'size': 10000}\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pip install biopython"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "kPuOqTOQdYvO",
        "outputId": "000ed4a5-0689-45dd-9eb3-e1c8e37d1100"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n",
            "Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (3.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m24.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.86\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import SeqIO\n",
        "from pathlib import Path"
      ],
      "metadata": {
        "id": "hMpUTRl7cXI8"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Make sure DATA_DIR exists\n",
        "DATA_DIR = Path(\"data/uniprot_sequences\")\n"
      ],
      "metadata": {
        "id": "sT9zzSzfcPvT"
      },
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "merged_fasta = DATA_DIR / \"all_viral_proteins.fasta\"\n"
      ],
      "metadata": {
        "id": "4rB3ESGlcgtZ"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "with open(merged_fasta, \"w\") as out_handle:\n",
        "    for fasta_file in DATA_DIR.glob(\"*.fasta\"):\n",
        "        if fasta_file.name != \"all_viral_proteins.fasta\":\n",
        "            for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
        "                SeqIO.write(record, out_handle, \"fasta\")"
      ],
      "metadata": {
        "id": "YG78d7G_cyTX"
      },
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(f\"[✓] Combined FASTA saved to {merged_fasta}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "uNDm5X5yc2iu",
        "outputId": "0a666215-ed3d-4775-e802-e5b403a617a5"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[✓] Combined FASTA saved to data/uniprot_sequences/all_viral_proteins.fasta\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "colab install -c bioconda hmmer\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 106
        },
        "id": "qXpNTA0khLwt",
        "outputId": "d67b2c93-8c7c-4d1a-c096-73c9f0af7ec6"
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "error",
          "ename": "SyntaxError",
          "evalue": "invalid syntax (ipython-input-1880986421.py, line 1)",
          "traceback": [
            "\u001b[0;36m  File \u001b[0;32m\"/tmp/ipython-input-1880986421.py\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    colab install -c bioconda hmmer\u001b[0m\n\u001b[0m          ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "sudo apt update\n",
        "sudo apt install hmmer\n",
        "wget http://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.hmm.gz\n",
        "gunzip Pfam-A.hmm.gz\n",
        "hmmpress Pfam-A.hmm\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 106
        },
        "id": "J9h8aEmzeCVV",
        "outputId": "b11a866a-ff6e-4bfe-e4f8-24933a61d95d"
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "error",
          "ename": "SyntaxError",
          "evalue": "invalid syntax (ipython-input-3372523686.py, line 1)",
          "traceback": [
            "\u001b[0;36m  File \u001b[0;32m\"/tmp/ipython-input-3372523686.py\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    sudo apt update\u001b[0m\n\u001b[0m         ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get update\n",
        "!apt-get install -y hmmer wget gzip\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "s4Gxw2KLhzHS",
        "outputId": "4c118483-388a-4cc7-a0c0-4e1c46aca693"
      },
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\r0% [Working]\r            \rGet:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]\n",
            "Get:2 https://cli.github.com/packages stable InRelease [3,917 B]\n",
            "Get:3 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]\n",
            "Get:4 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]\n",
            "Hit:5 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]\n",
            "Hit:7 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Get:8 https://r2u.stat.illinois.edu/ubuntu jammy/main all Packages [9,576 kB]\n",
            "Hit:9 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Get:10 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Packages [2,862 kB]\n",
            "Get:11 http://security.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [1,287 kB]\n",
            "Get:12 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]\n",
            "Get:13 http://archive.ubuntu.com/ubuntu jammy-updates/multiverse amd64 Packages [69.2 kB]\n",
            "Get:14 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,598 kB]\n",
            "Get:15 http://security.ubuntu.com/ubuntu jammy-security/restricted amd64 Packages [6,205 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [3,966 kB]\n",
            "Get:17 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [3,634 kB]\n",
            "Get:18 http://archive.ubuntu.com/ubuntu jammy-updates/restricted amd64 Packages [6,411 kB]\n",
            "Get:19 http://archive.ubuntu.com/ubuntu jammy-backports/main amd64 Packages [83.9 kB]\n",
            "Get:20 http://archive.ubuntu.com/ubuntu jammy-backports/universe amd64 Packages [37.2 kB]\n",
            "Fetched 36.1 MB in 2s (14.5 MB/s)\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "gzip is already the newest version (1.10-4ubuntu4.1).\n",
            "wget is already the newest version (1.21.2-2ubuntu1.1).\n",
            "The following additional packages will be installed:\n",
            "  libdivsufsort3\n",
            "Suggested packages:\n",
            "  hmmer-doc\n",
            "The following NEW packages will be installed:\n",
            "  hmmer libdivsufsort3\n",
            "0 upgraded, 2 newly installed, 0 to remove and 12 not upgraded.\n",
            "Need to get 1,198 kB of archives.\n",
            "After this operation, 7,621 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libdivsufsort3 amd64 2.0.1-5 [42.8 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 hmmer amd64 3.3.2+dfsg-1 [1,155 kB]\n",
            "Fetched 1,198 kB in 1s (1,137 kB/s)\n",
            "Selecting previously unselected package libdivsufsort3:amd64.\n",
            "(Reading database ... 117528 files and directories currently installed.)\n",
            "Preparing to unpack .../libdivsufsort3_2.0.1-5_amd64.deb ...\n",
            "Unpacking libdivsufsort3:amd64 (2.0.1-5) ...\n",
            "Selecting previously unselected package hmmer.\n",
            "Preparing to unpack .../hmmer_3.3.2+dfsg-1_amd64.deb ...\n",
            "Unpacking hmmer (3.3.2+dfsg-1) ...\n",
            "Setting up libdivsufsort3:amd64 (2.0.1-5) ...\n",
            "Setting up hmmer (3.3.2+dfsg-1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.11) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero_v2.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget http://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.hmm.gz\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "REJy_cX3h_d1",
        "outputId": "060b9035-d2cf-4719-8b4f-64b5e666507b"
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2026-01-07 12:56:50--  http://ftp.ebi.ac.uk/pub/databases/Pfam/current_release/Pfam-A.hmm.gz\n",
            "Resolving ftp.ebi.ac.uk (ftp.ebi.ac.uk)... 193.62.193.165\n",
            "Connecting to ftp.ebi.ac.uk (ftp.ebi.ac.uk)|193.62.193.165|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 355568131 (339M) [application/x-gzip]\n",
            "Saving to: ‘Pfam-A.hmm.gz’\n",
            "\n",
            "Pfam-A.hmm.gz       100%[===================>] 339.10M  37.6MB/s    in 9.3s    \n",
            "\n",
            "2026-01-07 12:57:00 (36.3 MB/s) - ‘Pfam-A.hmm.gz’ saved [355568131/355568131]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gunzip Pfam-A.hmm.gz\n"
      ],
      "metadata": {
        "id": "KPlsLWiqjC5q"
      },
      "execution_count": 23,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!hmmpress Pfam-A.hmm\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7ZGGZEU3jNuI",
        "outputId": "bbebc5de-91bd-4e0f-ef3b-921eee700a87"
      },
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Working...    done.\n",
            "Pressed and indexed 25545 HMMs (25545 names and 25545 accessions).\n",
            "Models pressed into binary file:   Pfam-A.hmm.h3m\n",
            "SSI index for binary model file:   Pfam-A.hmm.h3i\n",
            "Profiles (MSV part) pressed into:  Pfam-A.hmm.h3f\n",
            "Profiles (remainder) pressed into: Pfam-A.hmm.h3p\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!hmmscan -h\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1L7Omx6HjbIN",
        "outputId": "a4772054-f765-4632-d6b2-36b6bee6f986"
      },
      "execution_count": 25,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# hmmscan :: search sequence(s) against a profile database\n",
            "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
            "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
            "# Freely distributed under the BSD open source license.\n",
            "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
            "Usage: hmmscan [-options] <hmmdb> <seqfile>\n",
            "\n",
            "Basic options:\n",
            "  -h : show brief help on version and usage\n",
            "\n",
            "Options controlling output:\n",
            "  -o <f>           : direct output to file <f>, not stdout\n",
            "  --tblout <f>     : save parseable table of per-sequence hits to file <f>\n",
            "  --domtblout <f>  : save parseable table of per-domain hits to file <f>\n",
            "  --pfamtblout <f> : save table of hits and domains to file, in Pfam format <f>\n",
            "  --acc            : prefer accessions over names in output\n",
            "  --noali          : don't output alignments, so output is smaller\n",
            "  --notextw        : unlimit ASCII text output line width\n",
            "  --textw <n>      : set max width of ASCII text output lines  [120]  (n>=120)\n",
            "\n",
            "Options controlling reporting thresholds:\n",
            "  -E <x>     : report models <= this E-value threshold in output  [10.0]  (x>0)\n",
            "  -T <x>     : report models >= this score threshold in output\n",
            "  --domE <x> : report domains <= this E-value threshold in output  [10.0]  (x>0)\n",
            "  --domT <x> : report domains >= this score cutoff in output\n",
            "\n",
            "Options controlling inclusion (significance) thresholds:\n",
            "  --incE <x>    : consider models <= this E-value threshold as significant\n",
            "  --incT <x>    : consider models >= this score threshold as significant\n",
            "  --incdomE <x> : consider domains <= this E-value threshold as significant\n",
            "  --incdomT <x> : consider domains >= this score threshold as significant\n",
            "\n",
            "Options for model-specific thresholding:\n",
            "  --cut_ga : use profile's GA gathering cutoffs to set all thresholding\n",
            "  --cut_nc : use profile's NC noise cutoffs to set all thresholding\n",
            "  --cut_tc : use profile's TC trusted cutoffs to set all thresholding\n",
            "\n",
            "Options controlling acceleration heuristics:\n",
            "  --max    : Turn all heuristic filters off (less speed, more power)\n",
            "  --F1 <x> : MSV threshold: promote hits w/ P <= F1  [0.02]\n",
            "  --F2 <x> : Vit threshold: promote hits w/ P <= F2  [1e-3]\n",
            "  --F3 <x> : Fwd threshold: promote hits w/ P <= F3  [1e-5]\n",
            "  --nobias : turn off composition bias filter\n",
            "\n",
            "Other expert options:\n",
            "  --nonull2     : turn off biased composition score corrections\n",
            "  -Z <x>        : set # of comparisons done, for E-value calculation\n",
            "  --domZ <x>    : set # of significant seqs, for domain E-value calculation\n",
            "  --seed <n>    : set RNG seed to <n> (if 0: one-time arbitrary seed)  [42]\n",
            "  --qformat <s> : assert input <seqfile> is in format <s>: no autodetection\n",
            "  --cpu <n>     : number of parallel CPU workers to use for multithreads  [2]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import subprocess\n",
        "\n",
        "def run_phmmer(query_fasta, db_fasta, output_tbl):\n",
        "    cmd = [\"phmmer\", \"--tblout\", output_tbl, query_fasta, db_fasta]\n",
        "    subprocess.run(cmd)\n",
        "    print(f\"phmmer done: {output_tbl}\")\n",
        "\n",
        "# phmmer\n",
        "run_phmmer(\"HIV_Nef.fasta\", \"all_viral_proteins.fasta\", \"phmmer_Nef.tbl\")\n",
        "run_phmmer(\"HCV_Nef.fasta\", \"all_viral_proteins.fasta\", \"phmmer_Nef.tbl\")\n",
        "run_phmmer(\"SARS_COV_2_Nef.fasta\", \"all_viral_proteins.fasta\", \"phmmer_Nef.tbl\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Lis9I1kglswP",
        "outputId": "e16c1c62-5f16-4a64-fce8-6082315de00f"
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "phmmer done: phmmer_Nef.tbl\n",
            "phmmer done: phmmer_Nef.tbl\n",
            "phmmer done: phmmer_Nef.tbl\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def run_hmmscan(seqs_fasta, hmm_db, output_tbl):\n",
        "    cmd = [\"hmmscan\", \"--domtblout\", output_tbl, hmm_db, seqs_fasta]\n",
        "    subprocess.run(cmd)\n",
        "    print(f\"hmmscan done: {output_tbl}\")\n",
        "\n",
        "# hmmscan\n",
        "run_hmmscan(\"HIV_proteins.fasta\", \"Pfam-A.hmm\", \"hmmscan_HIV.tbl\")\n",
        "run_hmmscan(\"HCV_proteins.fasta\", \"Pfam-A.hmm\", \"hmmscan_HCV.tbl\")\n",
        "run_hmmscan(\"SARS_COV_2_proteins.fasta\", \"Pfam-A.hmm\", \"hmmscan_SARS_COV_2.tbl\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7pagHKTamC_D",
        "outputId": "1f32d64d-3825-4812-d69b-893bfa4efaeb"
      },
      "execution_count": 32,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "hmmscan done: hmmscan_HIV.tbl\n",
            "hmmscan done: hmmscan_HCV.tbl\n",
            "hmmscan done: hmmscan_SARS_COV_2.tbl\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def run_hmmsearch(hmm_profile, db_fasta, output_tbl):\n",
        "    cmd = [\"hmmsearch\", \"--tblout\", output_tbl, hmm_profile, db_fasta]\n",
        "    subprocess.run(cmd)\n",
        "    print(f\"hmmsearch done: {output_tbl}\")\n",
        "\n",
        "# hmmsearch\n",
        "run_hmmsearch(\"Nef.hmm\", \"all_viral_proteins.fasta\", \"hmmsearch_Nef.tbl\")\n",
        "run_hmmsearch(\"Nef.hmm\", \"all_viral_proteins.fasta\", \"hmmsearch_Nef.tbl\")\n",
        "run_hmmsearch(\"Nef.hmm\", \"all_viral_proteins.fasta\", \"hmmsearch_Nef.tbl\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lsJQOTe3mefd",
        "outputId": "1e27ae7c-d05c-4824-81bb-3371d9cb2cd0"
      },
      "execution_count": 33,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "hmmsearch done: hmmsearch_Nef.tbl\n",
            "hmmsearch done: hmmsearch_Nef.tbl\n",
            "hmmsearch done: hmmsearch_Nef.tbl\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def run_jackhmmer(query_fasta, db_fasta, output_tbl, iterations=3):\n",
        "    cmd = [\"jackhmmer\", \"--tblout\", output_tbl, \"-N\", str(iterations), query_fasta, db_fasta]\n",
        "    subprocess.run(cmd)\n",
        "    print(f\"jackhmmer done: {output_tbl}\")\n",
        "\n",
        "# Example\n",
        "run_jackhmmer(\"HIV_Nef.fasta\", \"all_viral_proteins.fasta\", \"jackhmmer_Nef.tbl\", iterations=3)\n",
        "run_jackhmmer(\"HCV_Nef.fasta\", \"all_viral_proteins.fasta\", \"jackhmmer_Nef.tbl\", iterations=3)\n",
        "run_jackhmmer(\"SARS_COV_2_Nef.fasta\", \"all_viral_proteins.fasta\", \"jackhmmer_Nef.tbl\", iterations=3)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pjfsJOh-neax",
        "outputId": "116ec4c9-cad9-499f-a052-749052378eb9"
      },
      "execution_count": 34,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "jackhmmer done: jackhmmer_Nef.tbl\n",
            "jackhmmer done: jackhmmer_Nef.tbl\n",
            "jackhmmer done: jackhmmer_Nef.tbl\n"
          ]
        }
      ]
    }
  ]
}